tgg/research/gnomad_tgps.py

```diff
@@ -0,0 +1,239 @@
+    ids = hl.set(ids)
+    return ids
+
+
+def write_output(
+        dat: hl.utils.struct.Struct,
+        file_dir: str,
+        filename: str,
+) -> None:
+    """
+    Write the output Struct containing computed stats to a .txt file.
+
+    :param hl.utils.struct.Struct dat: Data containing computed stats.
+    :param str file_dir: Output directory.
+    :param str filename: Output filename.
+    :return: None; function writes output to the output path.
+    """
+    with hfs.open(f'{file_dir}/{filename}', 'w') as f:
+        print(dat, file=f)
+
+
+def compute_stats(
+        pangenome_ids_path: str,
+        output_dir: str,
+) -> None:
+    """
+    Compute the stats of variants per gnomAD sample for gnomAD samples not in the
+    pangenome.
+
+    :param str pangenome_ids_path: Cloud path for the text file containing pangenome
+    sample IDs.
+    :param str output_dir: Cloud path to output the result file.
+    :return: None; function writes .txt file to output directory.
+    """
+    # Get variant site frequency table.
+    gnomad_freq_ht = release_sites().ht()
+
+    # Get gnomAD v3 data in split form.
+    mt = get_gnomad_v3_vds().variant_data
+    mt = hl.experimental.sparse_split_multi(mt)
+
+    # Get gnomAD v3 metadata and filter to released samples.
+    meta_ht = meta.ht()
+    meta_ht = meta_ht.filter(meta_ht.release == True)
+
+    # Use the filtered metadata to filter gnomAD v3 data.
+    mt = mt.filter_cols(hl.is_defined(meta_ht[mt.col_key]))
+    num_of_v3_samples = mt.count_cols()
+    logger.info("The number of samples in gnomAD v3 is: ", num_of_v3_samples)
+
+    # Get pangenome sample IDs.
+    pangenome_ids = get_pangenome_ids(pangenome_ids_path)
+    logger.info("The total number of samples in pangenome is ", hl.eval(hl.len(
+        pangenome_ids)))
+
+    # Filter mt to two new matrix tables: one containing gnomAD samples not in
+    # pangenome, and one containing samples in pangenome.
+    mt = mt.annotate_cols(s_in_pangenome=pangenome_ids.contains(mt.s))
+    pg_mt = mt.filter_cols(mt.s_in_pangenome)
+    mt = mt.filter_cols(~mt.s_in_pangenome)
+
+    # Log the numbers of gnomAD samples not in and in the pangenome reference.
+    num_in_samples = pg_mt.count_cols()
+    num_out_of_samples = mt.count_cols()
+    logger.info("The number of samples both in pangenome and gnomAD v3 is: ",
+                num_in_samples)
+    logger.info("The number of samples only in gnomAD v3 is: ", num_out_of_samples)
+
+    # Annotate the gnomAD matrix table with allele frequencies and
+    # filter out SNVs with minor allele frequencies lower than 1% and
+    # filter out samples with no variant call.
+    mt = mt.annotate_rows(freq=gnomad_freq_ht[mt.row_key].freq)
+    mt = mt.filter_rows(mt.freq.AF[1] > 0.01)  # ##why AF[1]?
+    mt = mt.filter_rows(hl.agg.any(mt.GT.is_non_ref()))
+
+    # Filter out samples with no variant call in the pangenome matrix table.
+    pg_mt = pg_mt.filter_rows(hl.agg.any(pg_mt.GT.is_non_ref()))
```

Comment:
this might make more logical sense if you group this command right after the `pg_mt` filter
```
mt = mt.filter_cols(~pangenome_ids.contains(mt.s))
# Filter to pangenome samples and their variants
pg_mt = hl.read_matrix_table("gs://gcp-public-data--gnomad/release/3.1.2/mt/genomes/gnomad.genomes.v3.1.2.hgdp_1kg_subset_sparse.mt")
pg_mt = pg_mt.filter_cols(pangenome_ids.contains(pg_mt.s))
pg_mt = pg_mt.filter_rows(hl.agg.any(pg_mt.GT.is_non_ref()))
```